homeassistant/components/shelly/__init__.py

```diff
@@ -142,6 +146,60 @@ async def async_device_setup(
         )
 
 
+async def async_services_setup(
+    hass: HomeAssistant, dev_reg: device_registry.DeviceRegistry
+):
+    """Set up services."""
+
+    async def async_service_ota_update(call: ServiceCall):
+        """Trigger OTA update."""
+        if not (call.data.get(ATTR_DEVICE_ID) or call.data.get(ATTR_AREA_ID)):
+            _LOGGER.warning("OTA update service: no device or area selected")
+            return
+
+        devices = []
+        if call.data.get(ATTR_AREA_ID):
+            for area_id in call.data.get(ATTR_AREA_ID):
+                devices += [
+                    area_dev
+                    for area_dev in device_registry.async_entries_for_area(
+                        dev_reg, area_id
+                    )
+                    if DOMAIN in next(iter(area_dev.identifiers))
+                ]
+
+        if call.data.get(ATTR_DEVICE_ID):
+            for device_id in call.data.get(ATTR_DEVICE_ID):
+                device = dev_reg.async_get(device_id)
+                if not any(device.id == x.id for x in devices):
+                    devices += [device]
+
+        for device in devices:
+            entry_id = next(iter(device.config_entries))
+            entry_data = hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry_id]
+            rest: ShellyDeviceRestWrapper = entry_data[REST]
```

Comment:
Note that `ShellyDeviceRestWrapper` is not created if the device is a sleeping device, the service should not rely on this wrapper, but on the main wrapper - `ShellyDeviceWrapper`
```suggestion
            device_wrapper: ShellyDeviceWrapper = entry_data[COAP]
```
see: https://github.com/home-assistant/core/blob/86176f1bf9ed90650693d608de4255bcbdb2f4a7/homeassistant/components/shelly/__init__.py#L133
In addition we should not try to trigger update for a sleeping devices, as I see it we have two options:
1. (preferred option) For a sleeping device, store a flag that we need to trigger an update and trigger an update upon CoAP event from the device, can be done here: https://github.com/home-assistant/core/blob/86176f1bf9ed90650693d608de4255bcbdb2f4a7/homeassistant/components/shelly/__init__.py#L184
2. For a sleeping device log a message that device is skipped since it is a sleeping device

Reply:
- the method is now moved into `ShellyDeviceRestWrapper` as `async_trigger_ota_update()`
- `async_trigger_ota_update()` uses the flag `self._ota_update_pending` for scheduling an ota update for sleeping devices